fifeforspark/lgb_modelers.py

```diff
@@ -0,0 +1,174 @@
+            subset = ~self.data[self.test_col] & ~self.data[self.predict_col]
+        data = self.label_data(time_horizon)
+        data = data.filter(subset)
+        data = self.subset_for_training_horizon(data, time_horizon)
+
+        train_data = data.filter(~data[self.validation_col])[
+            self.categorical_features + self.numeric_features
+            ]
+        indexers = [StringIndexer(inputCol=column, outputCol=column + "_index")
+                    for column in self.categorical_features]
+        feature_columns = [column + "_index" for column in self.categorical_features] + self.numeric_features
+        assembler = VectorAssembler(inputCols=feature_columns, outputCol='features')
+        lgb_model = lgb(featuresCol="features",
+                        labelCol="_label",
+                        *params[time_horizon],
+                        class_weight=data.filter(~data[self.validation_col])[self.weight_col]
+                        if self.weight_col
+                        else None
+                        )
+        pipeline = Pipeline(stages=[*indexers, assembler, lgb_model])
+        model = pipeline.fit(train_data)
+        return model
+
+    def predict(
+            self, subset: Union[None, pyspark.sql.column.Column] = None, cumulative: bool = True
+    ) -> pyspark.sql.DataFrame:
+        """Use trained LightGBM models to predict the outcome for each observation and time horizon.
+        Args:
+            subset: A Boolean Spark Column that is True for observations for which
+                predictions will be produced. If None, default to all
+                observations.
+            cumulative: If True, produce cumulative survival probabilies.
+                If False, produce marginal survival probabilities (i.e., one
+                minus the hazard rate).
+        Returns:
+            A dataframe of predictions by observation and lead
+            length.
+        """
+        subset = default_subset_to_all(subset, self.data)
+        predict_data = self.data.filter(subset)[self.categorical_features + self.numeric_features]
+
+        firstelement = udf(lambda v: float(v[0]), FloatType())
+        first_model = self.model[0]
+        predictions = first_model.transform(predict_data).selectExpr('probability as probability_1')
+        predictions = predictions.withColumn('probability_1', firstelement(predictions['probability_1']))
+        for i, lead_specific_model in enumerate(self.model):
+            if i != 0:
+                predictions = predictions.withColumn(f'probability_{i+1}',
+                                                     firstelement(predictions[f'probability_{i+1}']))
+                if cumulative:
+                    predictions = predictions.withColumn(f'probability_{i + 1}',
+                                                         predictions[f'probability_{i + 1}'] *
+                                                         predictions[f'probability_{i}'])
```

Comment:
I might be wrong, but it doesn’t seem like you’re calling .transform() on anything past the first model (no transforms on ‘lead_specific_model’ within the enumerate loop).